### ISSUE TITLE
Add third-part verification tests for AArch64 decoder/encoder

### ISSUE THREAD

**Issue by AssadHashmi** [DynamoRIO/dynamorio#4396]
The current test set for the AArch64 decoder/encoder checks internal consistency but not correctness against an external third-party reference decoder/encoder. For AArch64, GNU's binutils and/or QEMU could be used for such cross-checking.

The x86 decoder/encoder already does this using binutils and udis86.

This issue was raised as a result of discussions in #4395 , see https://github.com/DynamoRIO/dynamorio/pull/4395#discussion_r466503441.

**Comment by derekbruening** [comment]
Is the dis-a64 test today using binutils objdump in some fashion at least (I guess up to a human to notice if its operands are different from DR's though?)?

**Comment by AssadHashmi** [comment]
> Is the dis-a64 test today using binutils objdump in some fashion at least (I guess up to a human to notice if its operands are different from DR's though?)?

Yes, currently we use GCC+bunutils objdump in `dis-a64.txt` tests to establish that we decode to the correct opcode. We then re-encode from the IR signature (added as part of the test) to check the hex encoding created matches GCC's. This doesn't check the correctness of the signature in terms of instruction interpretation, just that the specified IR signature string generates the same hex encoding as was decoded.

The plan is to close the loop when we complete the encoder tests checking that the IR signature created by macros in `ir_aarch64.c` matches the signatures in `dis-a64.txt` generating the same hex encoding as GCC's in `dis-a64.txt`.

Is that adequate?

**Comment by derekbruening** [comment]
> Yes, currently we use GCC+bunutils objdump in dis-a64.txt tests to establish that we decode to the correct opcode.

The opcode matching is by a human when dis-a64.txt is updated?

> We then re-encode from the IR signature (added as part of the test) to check the hex encoding created matches GCC's. 

Not sure I follow "from the IR signature": do you mean the disassembly?  I believe the test decodes from the 4-byte hex encoding into IR and then encodes from IR to hex: so the disassembly strings in the file are not used by the test?

Also, from a glance at the test: I don't see the original raw bits being invalidated, so I would expect the re-encoding to just copy the original raw bits and not do any encoding from the IR?  Or does #5176 prevent this (I do see a check in instr_encode_arch() for the raw bits being valid turning into a memcpy but maybe they are never valid because of #5176)?  I would expect the test to explicitly mark the raw bits invalid to ensure full encoding is triggered.

> Is that adequate?

One measure might be whether it will prevent future errors of types we've seen before.  Looking at some past PR's that fix decoder errors: #5174 goes and expands what's in dis-a64.txt.  How are the new lines chosen?  Would those new lines have caught the bug that PR is fixing?  Similarly for some other past errors: #5086, #5047, #4395.

It seems the precise content of dis-a64.txt is a key factor in the coverage?  Is there a way to systematically check for coverage or automate generation of entries in that file?

An attempt at a simple test of every opcode being in there:
```
$ for i in $(egrep -o 'OP_[a-z_0-9]+ =' opcode_api.h | sed 's/OP_//;s/ =//;'); do grep -i -q $i ../src/suite/tests/api/dis-a64.txt || echo $i; done
bcond
fcmle
fcmlt
```
So it seems it is missing some opcodes?

It does seem that positive tests are definitely more important, but negative tests could be useful too: ensuring that encodings that don't map to any opcode aren't accidentally decoded as valid.

**Comment by AssadHashmi** [comment]
>> Yes, currently we use GCC+bunutils objdump in dis-a64.txt tests to establish that we decode to the correct opcode.

> The opcode matching is by a human when dis-a64.txt is updated?

No, we now generate instructions in `.s` assembly files, build with GCC, disassemble and use the disassembly in `dis-a64.txt` for hex encoding and DR IR. Previously it was human matching.

>> We then re-encode from the IR signature (added as part of the test) to check the hex encoding created matches GCC's.

> Not sure I follow "from the IR signature": do you mean the disassembly?

Yes.

> I believe the test decodes from the 4-byte hex encoding into IR and then encodes from IR to hex: so the disassembly strings in the file are not used by the test?

Sorry I don't understand. When encoding from IR to hex, isn't that the same as using the disassembly strings, i.e. IR to encode and then check the hex?

Perhaps this example will make what's happening clearer:
```
hex encoding : native assembly mnemonic      : IR
fa42c023     : ccmp   x1, x2, #0x3, gt       : ccmp.gt %x1 %x2 $0x03
```
The test first takes the hex and uses the decoder to check the correct IR is produced.
It then takes the IR and uses the encoder to generate the hex and checks it's the same.

So if there's an error in encoding the IR, this sort of thing happens:
```
219: Error: Reencoding differs:
219: fa42c023  ccmp.gt %x1 %x2 $0x03  <-- what the encoder should generate from IR
219: fa40c023  ccmp.gt %x1 %x1 $0x03  <-- what the encoder actually generates from IR
```

> Also, from a glance at the test: I don't see the original raw bits being invalidated, so I would expect the re-encoding to just copy the original raw bits and not do any encoding from the IR? Or does #5176 prevent this (I do see a check in instr_encode_arch() for the raw bits being valid turning into a memcpy but maybe they are never valid because of #5176)? I would expect the test to explicitly mark the raw bits invalid to ensure full encoding is triggered.

AFAICT `encode_common()` is called in `instr_encode_arch()` so it doesn't look like they are never valid. I'll do some more digging.

>> Is that adequate?

> One measure might be whether it will prevent future errors of types we've seen before. Looking at some past PR's that fix decoder errors: #5174 goes and expands what's in dis-a64.txt. How are the new lines chosen? Would those new lines have caught the bug that PR is fixing? Similarly for some other past errors: #5086, #5047, #4395.

> It seems the precise content of `dis-a64.txt` is a key factor in the coverage? Is there a way to systematically check for coverage or automate generation of entries in that file?

> So it seems it is missing some opcodes?

So, currently we're generating fairly basic tests just to make sure that obvious errors are caught. That's what the PRs you listed fixed. We are working on a more thorough test generator which is based on a machine readable spec. This is work in progress which we're planning to upstream and will include stress tests which we hope will flush out not just the obvious errors but corner cases as well. All opcodes and instruction variants will be tested by the generator.

> It does seem that positive tests are definitely more important, but negative tests could be useful too: ensuring that encodings that don't map to any opcode aren't accidentally decoded as valid.

Yes, that's something we need to address and the generator should help with that.

**Comment by derekbruening** [comment]
> > > Yes, currently we use GCC+bunutils objdump in dis-a64.txt tests to establish that we decode to the correct opcode.
> 
> > The opcode matching is by a human when dis-a64.txt is updated?
> 
> No, we now generate instructions in `.s` assembly files, build with GCC, disassemble and use the disassembly in `dis-a64.txt` for hex encoding and DR IR. Previously it was human matching.

OK so a robot is ensuring the DR disassembly column has the same opcode as the binutils disassembly column.

> > I believe the test decodes from the 4-byte hex encoding into IR and then encodes from IR to hex: so the disassembly strings in the file are not used by the test?
> 
> Sorry I don't understand. When encoding from IR to hex, isn't that the same as using the disassembly strings, i.e. IR to encode and then check the hex?
> 
> Perhaps this example will make what's happening clearer:
> 
> ```
> hex encoding : native assembly mnemonic      : IR
> fa42c023     : ccmp   x1, x2, #0x3, gt       : ccmp.gt %x1 %x2 $0x03
> ```
> 
> The test first takes the hex and uses the decoder to check the correct IR is produced. It then takes the IR and uses the encoder to generate the hex and checks it's the same.

OK, so it does in fact disassemble from DR IR to a string and compares to the DR IR string in the file.

> > Also, from a glance at the test: I don't see the original raw bits being invalidated, so I would expect the re-encoding to just copy the original raw bits and not do any encoding from the IR? Or does #5176 prevent this (I do see a check in instr_encode_arch() for the raw bits being valid turning into a memcpy but maybe they are never valid because of #5176)? I would expect the test to explicitly mark the raw bits invalid to ensure full encoding is triggered.
> 
> AFAICT `encode_common()` is called in `instr_encode_arch()` so it doesn't look like they are never valid. I'll do some more digging.

I assume you mean "it doesn't look like they are *ever* valid".  If the raw bits were valid, the test's call to encode would simply do a copy of the originally-decoded hex value and would not invoke the full encoder.  Hence why the test should invalidate the bits to avoid the fast-path "encode" of just copying bits.

> 
> > > Is that adequate?
> 
> > One measure might be whether it will prevent future errors of types we've seen before. Looking at some past PR's that fix decoder errors: #5174 goes and expands what's in dis-a64.txt. How are the new lines chosen? Would those new lines have caught the bug that PR is fixing? Similarly for some other past errors: #5086, #5047, #4395.
> 
> > It seems the precise content of `dis-a64.txt` is a key factor in the coverage? Is there a way to systematically check for coverage or automate generation of entries in that file?
> 
> > So it seems it is missing some opcodes?
> 
> So, currently we're generating fairly basic tests just to make sure that obvious errors are caught. That's what the PRs you listed fixed. We are working on a more thorough test generator which is based on a machine readable spec. This is work in progress which we're planning to upstream and will include stress tests which we hope will flush out not just the obvious errors but corner cases as well. All opcodes and instruction variants will be tested by the generator.

Sounds good!
